src/main/cocoa/CocoaWindow.mm

```diff
@@ -0,0 +1,883 @@
+                                        ue.nType       = UIE_HIDE;
+                                        handle_event(&ue);
+                                    }];
+
+                [center addObserverForName:NSWindowWillCloseNotification
+                                    object:window
+                                    queue:[NSOperationQueue mainQueue]
+                                    usingBlock:^(NSNotification *note) {
+                                        lsp_trace("UIE_HIDE");
+                                        event_t ue;
+                                        init_event(&ue);
+                                        ue.nType       = UIE_HIDE;
+                                        handle_event(&ue);
+                                    }];
+
+                //TODO: implement all notification events
+                /*
+                [center addObserverForName:NSWindowDidResizeNotification
+                                    object:window
+                                    queue:[NSOperationQueue mainQueue]
+                                    usingBlock:^(NSNotification *note) {
+                                        // Your UIE_HIDE handling here
+                                        //ue->nType = UIE_HIDE;
+                                        lsp_trace("UIE_RESIZE");
+                                        event_t ue;
+                                        init_event(&ue);
+                                        ue.nType       = UIE_RESIZE;
+                                        handle_event(&ue);
+                                    }]; */
+            }
+
+            void CocoaWindow::destroy()
+            {
+                
+                if (pCocoaDisplay != NULL)
+                    pCocoaDisplay->vWindows.qpremove(this);
+
+                // Surface related
+                if (pSurface != NULL)
+                {
+                    pSurface->destroy();
+                    delete pSurface;
+                    pSurface = NULL;
+                }
+                /*
+                if (!bWrapper)
+                    DestroyWindow(hWindow);
+                */
+                //pCocoaDisplay = NULL;
+
+                IWindow::destroy();
+            }
+
+            NSWindow *CocoaWindow::nswindow() const {
+                return pCocoaWindow;
+            }
+
+            ISurface *CocoaWindow::create_surface(CocoaDisplay *display, size_t width, size_t height)
+            {
+                ISurface *result = NULL;
+
+                if (result == NULL)
+                {
+                    
+                    result = new CocoaCairoSurface(display, width, height);
+                    if (result != NULL)
+                        lsp_trace("Using CocoaCairoSurface ptr=%p", result);
+                }
+                return result;
+            }
+
+            // Triggers redraw from app and tests
+            status_t CocoaWindow::invalidate()
```

Comment:
I think the problem is here. The `invalidate` call triggers the window system to mark the window content being invalid and need for redraw. 

For Windows, we use `InvalidateRect` to make the window handler to receive the redraw event:

https://github.com/lsp-plugins/lsp-ws-lib/blob/d4b9f0525a2df969d87d269b3cc90c7dc707348d/src/main/win/WinWindow.cpp#L863-L870

For X11 we simulate window invalidation by sending an `ExposeEvent`:

https://github.com/lsp-plugins/lsp-ws-lib/blob/d4b9f0525a2df969d87d269b3cc90c7dc707348d/src/main/x11/X11Window.cpp#L500-L522

But in your implementation you just call the window to become redrawn, and this is not correct behaviour.